File: Binary Search/median_matrix_rowsorted.py


"""
    def median(matrix: [[int]], m: int, n: int) -> int:
        # Write your code here.
        arr = []
        for i in range(m):
            for j in range(n):
                arr.append(matrix[i][j])
        arr.sort()
        return arr[len(arr)//2]
        pass
"""

from bisect import bisect_right

def median(matrix, m, n):

    low = min(row[0] for row in matrix)
    high = max(row[-1] for row in matrix)

    req = (m * n) // 2

    while low <= high:
        mid = (low + high) // 2

        count = 0
        for row in matrix:
            count += bisect_right(row, mid)

        if count <= req:
            low = mid + 1
        else:
            high = mid - 1

    return low

File: Binary Search/median_matrix_rowsorted.java
class Solution {
    public int upperBound(int[][] mat, int x, int m, int n){
        int cnt = 0;
        for(int i=0;i<m;i++){
            int low = 0;
            int high = n;
            int[] arr = mat[i];
            while(low<high){
                int mid = (low+high)/2;
                if(arr[mid]<=x){
                    low = mid + 1;
                } else {
                    high = mid;
                }
            }
            cnt += low;
        }
        return cnt;
    }
    public int median(int[][] mat) {
        // code here
        int m = mat.length;
        int n = mat[0].length;
        int low = Integer.MAX_VALUE;
        int high = Integer.MIN_VALUE;
        
        for(int i =0;i<m;i++){
            low = Math.min(low, mat[i][0]);
            high = Math.max(high,mat[i][n-1]);
        }
        int req = (m*n)/2;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(upperBound(mat,mid,m,n) <= req){
                low = mid+1;
            } else{
                high = mid - 1;
            }
        }
        return low;
        
    }
}